File: backend/app/services/claude_service.py
import json

import anthropic

from app.config import settings

SUMMARIZE_SYSTEM_PROMPT = (
    "You are a news curator for a daily briefing app called Daily Kernel. "
    "Summarize each article into a compelling 2-3 sentence summary. "
    "Focus on why this matters and what's new. Be concise and engaging."
)


async def summarize_articles(
    category_name: str, articles: list[dict]
) -> list[dict]:
    """Summarize a list of articles using Claude.

    Each article dict should have: title, url, snippet, source_name.
    Returns a list of dicts with: title, summary, source_url, source_name.
    """
    if not articles:
        return []

    articles_text = ""
    for i, article in enumerate(articles, 1):
        articles_text += (
            f"Article {i}:\n"
            f"Title: {article['title']}\n"
            f"Source: {article.get('source_name', 'Unknown')}\n"
            f"Snippet: {article.get('snippet', 'No preview available')}\n\n"
        )

    user_prompt = (
        f"Category: {category_name}\n\n"
        f"{articles_text}"
        f"For each article, provide a JSON array of objects with these fields:\n"
        f'- "title": the article title (keep original or improve slightly)\n'
        f'- "summary": a 2-3 sentence summary\n'
        f'- "article_index": the 1-based index of the article\n\n'
        f"Respond ONLY with the JSON array, no other text."
    )

    client = anthropic.AsyncAnthropic(api_key=settings.ANTHROPIC_API_KEY)

    try:
        message = await client.messages.create(
            model=settings.CLAUDE_MODEL,
            max_tokens=2048,
            system=SUMMARIZE_SYSTEM_PROMPT,
            messages=[{"role": "user", "content": user_prompt}],
        )

        response_text = message.content[0].text.strip()

        if response_text.startswith("```"):
            lines = response_text.split("\n")
            lines = lines[1:]
            if lines and lines[-1].strip() == "```":
                lines = lines[:-1]
            response_text = "\n".join(lines)

        summaries_raw = json.loads(response_text)

        results = []
        for item in summaries_raw:
            idx = item.get("article_index", 0) - 1
            source_article = articles[idx] if 0 <= idx < len(articles) else {}

            results.append(
                {
                    "title": item.get("title", source_article.get("title", "")),
                    "summary": item.get("summary", ""),
                    "source_url": source_article.get("url", ""),
                    "source_name": source_article.get("source_name", ""),
                }
            )

        return results

    except (json.JSONDecodeError, KeyError, IndexError):
        return [
            {
                "title": article["title"],
                "summary": article.get("snippet", "Summary unavailable."),
                "source_url": article.get("url", ""),
                "source_name": article.get("source_name", ""),
            }
            for article in articles
        ]
    except anthropic.APIError:
        return [
            {
                "title": article["title"],
                "summary": article.get("snippet", "Summary unavailable."),
                "source_url": article.get("url", ""),
                "source_name": article.get("source_name", ""),
            }
            for article in articles
        ]


async def analyze_feedback(feedback_data: list[dict]) -> str:
    """Analyze user feedback patterns and return insights.

    feedback_data: list of dicts with keys: category_name, action, count.
    Returns a string with insights about user preferences.
    """
    if not feedback_data:
        return "No feedback data available yet."

    feedback_text = "User feedback summary:\n"
    for item in feedback_data:
        feedback_text += (
            f"- {item['category_name']}: "
            f"{item.get('thumbs_up', 0)} thumbs up, "
            f"{item.get('thumbs_down', 0)} thumbs down, "
            f"{item.get('skips', 0)} skips\n"
        )

    client = anthropic.AsyncAnthropic(api_key=settings.ANTHROPIC_API_KEY)

    try:
        message = await client.messages.create(
            model=settings.CLAUDE_MODEL,
            max_tokens=512,
            system=(
                "You are an analytics assistant for Daily Kernel, a news briefing app. "
                "Analyze the user's feedback patterns and provide brief, actionable insights "
                "about their content preferences. Be concise - 2-3 sentences max."
            ),
            messages=[
                {
                    "role": "user",
                    "content": f"Analyze this feedback data and suggest how to improve the briefing:\n\n{feedback_text}",
                }
            ],
        )
        return message.content[0].text.strip()
    except anthropic.APIError:
        return "Unable to analyze feedback at this time."

File: src/lib/claude-service.ts
import Anthropic from '@anthropic-ai/sdk';
import { NewsArticle, CardSummary } from '@/types';

const CLAUDE_MODEL = 'claude-sonnet-4-20250514';

function getClient(): Anthropic {
  return new Anthropic({
    apiKey: process.env.ANTHROPIC_API_KEY!,
  });
}

/**
 * Summarize a batch of news articles for a given category into briefing cards.
 * Returns an array of card summaries with title, summary, source URL, and source name.
 */
export async function summarizeArticles(
  categoryName: string,
  articles: NewsArticle[]
): Promise<CardSummary[]> {
  if (articles.length === 0) {
    return [];
  }

  const client = getClient();

  const articlesText = articles
    .map(
      (a, i) =>
        `Article ${i + 1}:\nTitle: ${a.title}\nSource: ${a.source_name}\nURL: ${a.url}\nSnippet: ${a.snippet}`
    )
    .join('\n\n');

  const systemPrompt = `You are a news curator for a daily briefing app called Daily Kernel. Your job is to create concise, informative card summaries from news articles. Each card should give the reader a quick understanding of the key story in 2-3 sentences. Be factual, neutral, and focus on what matters most. Avoid sensationalism. If multiple articles cover the same story, consolidate them into one card and pick the best source.`;

  const userPrompt = `Category: ${categoryName}

Here are the articles to summarize:

${articlesText}

Create a JSON array of card summaries. Each card should have:
- "title": A clear, concise headline (max 80 chars)
- "summary": A 2-3 sentence summary of the key points
- "source_url": The URL of the primary source article
- "source_name": The name of the source

Respond ONLY with a valid JSON array. No other text.`;

  try {
    const message = await client.messages.create({
      model: CLAUDE_MODEL,
      max_tokens: 2048,
      messages: [
        {
          role: 'user',
          content: userPrompt,
        },
      ],
      system: systemPrompt,
    });

    const responseText =
      message.content[0].type === 'text' ? message.content[0].text : '';

    return parseCardSummaries(responseText, articles);
  } catch (error) {
    console.error('Claude summarization error:', error);
    // Fallback: use raw article snippets as summaries
    return fallbackSummaries(articles);
  }
}

/**
 * Parse Claude's JSON response into card summaries, with fallback handling.
 */
function parseCardSummaries(
  responseText: string,
  originalArticles: NewsArticle[]
): CardSummary[] {
  try {
    // Try to extract JSON from the response (handle markdown code blocks)
    let jsonStr = responseText.trim();
    const jsonMatch = jsonStr.match(/```(?:json)?\s*([\s\S]*?)```/);
    if (jsonMatch) {
      jsonStr = jsonMatch[1].trim();
    }

    const parsed = JSON.parse(jsonStr);

    if (!Array.isArray(parsed)) {
      throw new Error('Response is not an array');
    }

    return parsed.map((item: Record<string, unknown>) => ({
      title: String(item.title || 'Untitled'),
      summary: String(item.summary || ''),
      source_url: item.source_url ? String(item.source_url) : null,
      source_name: item.source_name ? String(item.source_name) : null,
    }));
  } catch (parseError) {
    console.error('Failed to parse Claude response:', parseError);
    return fallbackSummaries(originalArticles);
  }
}

/**
 * Fallback: convert raw articles into card summaries when Claude is unavailable.
 */
function fallbackSummaries(articles: NewsArticle[]): CardSummary[] {
  return articles.map((article) => ({
    title: article.title,
    summary: article.snippet || 'No summary available.',
    source_url: article.url || null,
    source_name: article.source_name || null,
  }));
}

/**
 * Analyze user feedback data and generate insights about reading preferences.
 * Returns a human-readable string with insights.
 */
export async function analyzeFeedback(
  feedbackData: {
    category: string;
    thumbs_up: number;
    thumbs_down: number;
    skip: number;
  }[]
): Promise<string> {
  if (feedbackData.length === 0) {
    return 'No feedback data available to analyze.';
  }

  const client = getClient();

  const dataText = feedbackData
    .map(
      (f) =>
        `${f.category}: ${f.thumbs_up} liked, ${f.thumbs_down} disliked, ${f.skip} skipped`
    )
    .join('\n');

  const prompt = `Here is a user's feedback data from a daily news briefing app:

${dataText}

Based on this feedback, provide 2-3 brief, actionable insights about their reading preferences. What categories do they engage with most? What should be adjusted? Keep it concise and friendly.`;

  try {
    const message = await client.messages.create({
      model: CLAUDE_MODEL,
      max_tokens: 512,
      messages: [
        {
          role: 'user',
          content: prompt,
        },
      ],
      system:
        'You are a helpful assistant that analyzes reading habits and provides brief, actionable insights.',
    });

    return message.content[0].type === 'text'
      ? message.content[0].text
      : 'Unable to generate insights.';
  } catch (error) {
    console.error('Claude feedback analysis error:', error);
    return 'Unable to analyze feedback at this time. Please try again later.';
  }
}
